File: 08/fall/tumble_down.py
from sys import stdin 

def top((x, y, w, h)): return y + h

def bottom((x, y, w, h)): return y

def left((x, y, w, h)): return x

def right((x, y, w, h)): return x + w

def center((x, y, w, h)): return x + w/2.0


def left_supports(a, b):
    return bottom(b) == top(a) and left(b) < right(a) and center(b) > left(a)

def right_supports(a, b):
    return bottom(b) == top(a) and right(b) > left(a) and center(b) < right(a)

def find_supported(blocks):
    supported = set()
    for b in blocks:
        if bottom(b) == 0:
            supported.add(b)
    changed = True
    while changed:
        to_add = set()
        changed = False
        for b in set(blocks) - supported:
            for a_l in supported:
                for a_r in supported:
                    if left_supports(a_l, b) and right_supports(a_r, b):
                        to_add.add(b)
                        changed = True
        supported = supported.union(to_add)
    return list(supported)

if __name__ == '__main__':
    blocks = [tuple(map(int, block.split())) for block in stdin.readlines()]
    supported = find_supported(blocks)
    for i, block in enumerate(blocks):
        if block in supported:
            print "block %d is supported" % (i+1)
        else:
            print "block %d is not supported" % (i + 1)

File: 08/fall/solutions/tumble_down.java
import java.util.*;

class tumble_down {

	private static int n = 0;

	private static class Block implements Comparable<Block> {

		int x, y, w, h;
		int number;
		boolean supported;

		Block(int x, int y, int w, int h) {
			this.x = x;
			this.y = y;
			this.w = w;
			this.h = h;
			this.number = ++n;
			this.supported = bottom() == 0;
		}

		int bottom() {return y;}
		int top() {return y+h;}
		int left() {return x;}
		int right() {return x+w;}
		double center() {return x+w/2.;}
		boolean leftSupports(Block b) {return b.bottom() == top() && b.left() < right() && b.center() > left();}
		boolean rightSupports(Block b) {return b.bottom() == top() && b.right() > left() && b.center() < right();}

		public int compareTo(Block to) {
			if (bottom() != to.bottom()) return bottom() - to.bottom();
			return number - to.number;
		}
	}

	public static void main(String[] arguments) {

		Scanner in = new Scanner(System.in);
		List<Block> blocks = new ArrayList<Block>();
		SortedSet<Block> sorted_blocks = new TreeSet<Block>();

		while (in.hasNextInt()) {
			Block b = new Block(in.nextInt(), in.nextInt(), in.nextInt(), in.nextInt());
			blocks.add(b);
			sorted_blocks.add(b);
		}
		for (Block over : sorted_blocks) {

			if (over.supported) continue; 
	
			boolean leftSupported = false, rightSupported = false;

			for (Block under : blocks) {
				if (! under.supported) continue;
				if (under.leftSupports(over)) leftSupported = true;
				if (under.rightSupports(over)) rightSupported = true;
			}
			if (leftSupported && rightSupported) over.supported = true;
		}
		for (Block b : blocks)
			System.out.printf("block %d is %ssupported%n", b.number, b.supported ? "" : "not ");
	}
}
